### PR TITLE
NAS-102286 / 11.3 / Multiple Repository Support For Plugins

### PR DESCRIPTION
**Pull request by sonicaj**
This PR introduces following changes making sure we support multiple repositories:

1) New Service to manage plugins i.e creating, listing installed/available etc.
2) New API for listing branches of a github repository.
3) New API for listing available/downloaded releases.
4) New Plugin service now allows multiple repository support.
5) Jail names are now expected when creating a plugin and the api/iocage will not do it automatically from middlewared's end.
6) Backward compatibility with jail.fetch method has been introduced which means plugins can still be created with jail.fetch but if multiple repository support is required, new plugin service should be used.

It does not introduce new api for listing templates as that can already be done by querying jail service with a filter.

### CODE REVIEW

File: src/middlewared/middlewared/plugins/jail.py [hunk 1]
```diff
@@ -117,6 +117,168 @@ def query(self, filters=None, options=None):
 
         return filter_list(jails, filters, options)
 
+    @accepts()
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/jail.py]: It almost seems like we need a new service called "plugin" to separate the logic?

Could be in the same file.

Thoughts?

Comment (sonicaj) [reply src/middlewared/middlewared/plugins/jail.py]: @william-gr i couldn't agree more. I was afraid of breaking the API too much this way. `fetch` too is overloaded and having a new service will help us manage that too in a better way. Please advise if you think i should move ahead with this and update the PR.

Comment (william-gr) [reply src/middlewared/middlewared/plugins/jail.py]: Yes, lets go ahead. We can try to somehow keep backward compatibility if at all possible, but currently there is too much confusion in the API with jails vs plugins

File: src/middlewared/middlewared/plugins/jail.py [hunk 2]
```diff
@@ -36,6 +36,430 @@
...
+    @accepts(
+        Dict(
+            'plugin_create',
+            Str('plugin_name', required=True),
+            Str('jail_name', required=True),
+            List('props', default=[], empty=False),
+            Str('branch', default=None, null=True),
+            Str('plugin_repository', default='https://github.com/freenas/iocage-ix-plugins.git'),
+        )
+    )
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/jail.py]: `@job` doesnt work here? It should

File: src/middlewared/middlewared/plugins/jail.py [hunk 3]
```diff
@@ -36,6 +36,430 @@
...
+    @job(
+        lock=lambda args: 'available_plugins_{}_{}'.format(
+            args[-1].get('branch') if args else None,
+            args[-1]['plugin_repository'] if args and args[-1].get(
+                'plugin_repository'
+            ) else 'https://github.com/freenas/iocage-ix-plugins.git'
+        )
+    )
+    def available_plugins(self, job, options):
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/jail.py]: Just `available` would be OK I think.

We are already in plugins namespace.

File: src/middlewared/middlewared/plugins/jail.py [hunk 4]
```diff
@@ -117,6 +580,45 @@ def query(self, filters=None, options=None):
...
+    @accepts(
+        Dict(
+            'branches_choices',
+            Str('account', default='freenas'),
+            Str('repository', default='iocage-ix-plugins'),
+        )
+    )
+    def branches_choices(self, options):
+        """
+        List branches for a github `account`'s `repository`.
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/jail.py]: Not a fan of relying on github for that. Is there anything we can do about that?

File: src/middlewared/middlewared/plugins/jail.py [hunk 5]
```diff
@@ -36,6 +36,469 @@
...
+            try:
+                for mac in value.split():
+                    MACAddr()(mac)
+
+                if len(value.split()) != 2 or any(value.split().count(v) > 1 for v in value.split()):
+                    raise ValueError('Exception')
+            except ValueError:
+                verrors.add(
+                    f'{schema}.{key}',
+                    'Please Enter two valid and different '
```
Comment (Mrt134) [inline src/middlewared/middlewared/plugins/jail.py]:
```suggestion
                    'Please enter two valid and different '
```

File: src/middlewared/middlewared/plugins/jail.py [hunk 6]
```diff
@@ -36,6 +36,469 @@
...
+            List('props', default=[], empty=False),
+            Str('branch', default=None, null=True),
+            Str('plugin_repository', default='https://github.com/freenas/iocage-ix-plugins.git'),
+        )
+    )
+    def do_create(self, data):
+        """
+        Create a Plugin.
+
+        `plugin_name` is the name of the plugin as specified by the INDEX file in "plugin_repository" and it's json
```
Comment (Mrt134) [inline src/middlewared/middlewared/plugins/jail.py]:
```suggestion
        `plugin_name` is the name of the plugin specified by the INDEX file in "plugin_repository" and its JSON
```

File: src/middlewared/middlewared/plugins/jail.py [hunk 7]
```diff
@@ -36,6 +36,469 @@
...
+        file.
+
+        `jail_name` is the name of the plugin jail and is required.
```
Comment (Mrt134) [inline src/middlewared/middlewared/plugins/jail.py]:
```suggestion
        `jail_name` is the name of the jail that will manage the plugin. Required.
```

File: src/middlewared/middlewared/plugins/jail.py [hunk 8]
```diff
@@ -36,6 +36,469 @@
...
+
+        `props` is a list of iocage props which user can manually set. For plugins, it should be ensured that the
```
Comment (Mrt134) [inline src/middlewared/middlewared/plugins/jail.py]:
```suggestion
        `props` is a list of jail properties that the user manually sets. Plugins should always set the
```

File: src/middlewared/middlewared/plugins/jail.py [hunk 9]
```diff
@@ -36,6 +36,469 @@
...
+        jails have networking capability set either with dhcp, ip address or nat.
```
Comment (Mrt134) [inline src/middlewared/middlewared/plugins/jail.py]:
```suggestion
        jail networking capability with DHCP, IP Address, or NAT properties.
```
Might be helpful to have an example (dhcp=on, ip4_addr="em0|192.168.0.1/24", etc) or link to iocage manpage?

File: src/middlewared/middlewared/plugins/jail.py [hunk 10]
```diff
@@ -36,6 +36,469 @@
...
+
+        `plugin_repository` is a git uri which will be used to fetch `plugin_name`'s data.
```
Comment (Mrt134) [inline src/middlewared/middlewared/plugins/jail.py]:
```suggestion
        `plugin_repository` is a git URI that fetches data for `plugin_name`.
```

File: src/middlewared/middlewared/plugins/jail.py [hunk 11]
```diff
@@ -36,6 +36,469 @@
...
+
+        `branch` is the branch to use for `plugin_repository`. It defaults to system's version. i.e 11.3-RELEASE.
```
Comment (Mrt134) [inline src/middlewared/middlewared/plugins/jail.py]:
```suggestion
        `branch` is the FreeNAS repository branch to use as the base for the `plugin_repository`. The default is to use the current system version. Examples: 11.3-RELEASE, 11.3-U1, 11.3-U2.
```
Suggested explaining the term and behavior a little more.

File: src/middlewared/middlewared/plugins/jail.py [hunk 12]
```diff
@@ -24,18 +24,483 @@
...
+        """
+        Update plugin `jail`.
+        """
+        await self._get_instance(jail)
+        return await self.middleware.call('jail.update', jail, data)
+
+    @accepts(Str('id'))
+    async def do_delete(self, id):
+        """
+        Delete plugin `jail`.
```
[the remark this answers is not in the document]
Comment (sonicaj) [reply src/middlewared/middlewared/plugins/jail.py]: Updated :)

**Pull request merged**